src/compiler/checker.ts

```diff
@@ -6551,19 +6609,30 @@ module ts {
             }
         }
 
+        function createPromiseType(awaitedType: Type, func: FunctionLikeDeclaration): Type {
+            if (globalPromiseSymbol) {
+                awaitedType = getAwaitedType(awaitedType, awaitedType);
+                var promiseType = globalPromiseType || getDeclaredTypeOfSymbol(globalPromiseSymbol);
+                return promiseType !== emptyObjectType ? createTypeReference(<GenericType>promiseType, [awaitedType]) : emptyObjectType;
+            }
+            error(func, Diagnostics.An_async_function_or_method_must_have_a_valid_awaitable_return_type);
+            return unknownType;
+        }

```

Comment:
Yes, in `checkCollisionWithAwaiterVariablesInGeneratedCode`.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Thursday, February 12, 2015 3:54 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/checker.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24633207>:

> @@ -6551,19 +6609,30 @@ module ts {

>              }

>          }

>

> +        function createPromiseType(awaitedType: Type, func: FunctionLikeDeclaration): Type {

> +            if (globalPromiseSymbol) {

> +                awaitedType = getAwaitedType(awaitedType, awaitedType);

> +                var promiseType = globalPromiseType || getDeclaredTypeOfSymbol(globalPromiseSymbol);

> +                return promiseType !== emptyObjectType ? createTypeReference(<GenericType>promiseType, [awaitedType]) : emptyObjectType;

> +            }

> +            error(func, Diagnostics.An_async_function_or_method_must_have_a_valid_awaitable_return_type);

> +            return unknownType;

> +        }

I assume elsewhere, you will check that Promise indeed resolves in the value space to the globalPromiseSymbol

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24633207>.